### PR TITLE
runtime: ENVOY_BUG for unknown flags

### PR DESCRIPTION
**Pull request by alyssawilk**
I posit we should do this in 2 phases  - envoy bug for one release, and reject config for the next.

Risk Level: Medium
Testing: new tests
Docs Changes: inline
Release Notes: inline
Part of https://github.com/envoyproxy/envoy/issues/20212

### CODE REVIEW

File: docs/root/version_history/current.rst [hunk 1]
```diff
@@ -27,6 +27,7 @@ Minor Behavior Changes
 * listener: the :ref:`ipv4_compat <envoy_api_field_core.SocketAddress.ipv4_compat>` flag can only be set on Ipv6 address and Ipv4-mapped Ipv6 address. A runtime guard is added ``envoy.reloadable_features.strict_check_on_ipv4_compat`` and the default is true.
 * perf: ssl contexts are now tracked without scan based garbage collection and greatly improved the performance on secret update.
 * router: record upstream request timeouts for all the cases and not just for those requests which are awaiting headers. This behavioral change can be temporarily reverted by setting runtime guard ``envoy.reloadable_features.do_not_await_headers_on_upstream_timeout_to_emit_stats`` to false.
+* runtime: setting deprecated runtime flags to ENVOY_BUG, rather than silently resulting in unexpected behavior on the data plane by no longer applying removed code paths. In the next version of Envoy, Envoy will reject such buggy configuration.
```
Comment (daixiang0) [inline docs/root/version_history/current.rst:30]: It is not really BUG, could we have a better name?

Comment (alyssawilk) [reply docs/root/version_history/current.rst:30]: generally using removed runtime guards is a bug in config, so I think this applies :-)

Comment (daixiang0) [reply docs/root/version_history/current.rst:30]: Oh, I see.

Comment (htuch) [reply docs/root/version_history/current.rst:30]: A couple of things:
1. As currently implemented, I think this just covers static and RTDS layers, which is probably the most important. I think admin overrides won't be caught, but this is probably fine (but maybe call it out?).
2. Would be good to close on the promise of rejecting this configuration.

File: source/common/runtime/runtime_features.cc [hunk 2]
```diff
@@ -82,6 +82,12 @@ ABSL_FLAG(uint64_t, re2_max_program_size_warn_level,            // NOLINT
 namespace Envoy {
 namespace Runtime {
 
+bool hasRuntimePrefix(absl::string_view feature) {
+  return (absl::StartsWith(feature, "envoy.reloadable_features.") &&
+          !absl::StartsWith(feature, "envoy.reloadable_features.FLAGS_quic")) ||
```
Comment (htuch) [inline source/common/runtime/runtime_features.cc:87]: Why this carve out?

**Review by htuch — COMMENTED** [review]
Looks good, just a couple of comments.

File: source/common/runtime/runtime_features.cc (hunk 2 above)
Comment (alyssawilk) [reply source/common/runtime/runtime_features.cc:89]: QUICHE flags aren't a part of the runtime flag framework below so need to be exempted.  commented in code.

**Review by htuch — APPROVED** [review]
LGTM, thanks!

**Comment by htuch** [comment]
Needs merge with main. Feel free to do the PR merge yourself once that is done.

**Comment by alyssawilk** [comment]
unfortunately needs a stamp for merge conflicts. PTAL?

**Review by htuch — APPROVED** [review]

**Pull request merged**